File: sctype_score/execution.R
#' sctype_score — ScType Marker Gene Scoring for Cell Type Identification
#'
#' PURPOSE:
#'   Automatically assigns cell types to clusters using marker gene enrichment
#'   scoring. Uses either built-in tissue-specific marker databases or custom
#'   XLSX files. Requires scaled expression data.
#'
#' INPUTS:
#'   input_data  : character (RDS path) | Seurat object
#'   params_list : list (optional, overrides defaults)
#'
#' DEFAULTS:
#'   tissue_type = "Immune system", db_path = NULL, scaled_assay = "scale.data",
#'   cluster_col = "seurat_clusters", score_col = "sctype_classification"
#'
#' OUTPUTS:
#'   Seurat object with ScType classifications in @meta.data
#'   and @misc$analysis_history updated

# ScType source functions (loaded from GitHub or local cache)
.load_sctype <- function() {
    # Try to load from installed package first
    if (requireNamespace("sctype", quietly = TRUE)) {
        return(invisible(NULL))
    }
    # Load core functions from GitHub
    sctype_score_url  <- "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R"
    gene_sets_url     <- "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R"
    for (url in c(sctype_score_url, gene_sets_url)) {
        source(url)
    }
}

run_sctype_score <- function(input_data, params_list = NULL) {
    # Memory-First
    if (is.character(input_data)) {
        obj <- readRDS(input_data)
    } else {
        obj <- input_data
    }

    # Parameter Safety
    default_params <- list(
        tissue_type    = "Immune system",
        db_path        = NULL,
        scaled_assay   = "scale.data",
        cluster_col    = "seurat_clusters",
        score_col      = "sctype_classification",
        top_cell_types = 1L
    )
    current_params <- modifyList(
        default_params,
        if (!is.null(params_list)) params_list else list()
    )

    # Load ScType functions
    .load_sctype()

    # Resolve marker database
    db <- if (!is.null(current_params$db_path)) {
        current_params$db_path
    } else {
        # Built-in ScType database from GitHub
        "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx"
    }

    # Prepare gene sets
    gs_list <- gene_sets_prepare(db, current_params$tissue_type)

    # Get scaled expression matrix
    assay_name <- current_params$scaled_assay
    if (assay_name == "scale.data") {
        scaled_mat <- Seurat::GetAssayData(obj, slot = "scale.data")
    } else if (assay_name == "SCT") {
        scaled_mat <- Seurat::GetAssayData(obj, assay = "SCT", slot = "scale.data")
    } else {
        scaled_mat <- Seurat::GetAssayData(obj, assay = assay_name, slot = "scale.data")
    }

    if (nrow(scaled_mat) == 0) {
        stop(paste0(
            "Scaled data not found in assay '", assay_name, "'. ",
            "Run ScaleData() or SCTransform() first."
        ))
    }

    # Compute ScType scores
    es_max <- sctype_score(
        scRNAseqData = scaled_mat,
        scaled       = TRUE,
        gs           = gs_list$gs_positive,
        gs2          = gs_list$gs_negative
    )

    # Assign cell types per cluster
    cluster_col <- current_params$cluster_col
    if (!cluster_col %in% colnames(obj@meta.data)) {
        # Fallback to available cluster column
        for (alt in c("seurat_clusters", "leiden", "louvain")) {
            if (alt %in% colnames(obj@meta.data)) {
                cluster_col <- alt
                break
            }
        }
    }

    clusters   <- obj@meta.data[[cluster_col]]
    cL_results <- do.call("rbind", lapply(unique(clusters), function(cl) {
        es_cl <- sort(rowSums(es_max[, rownames(obj@meta.data)[clusters == cl], drop = FALSE]),
                      decreasing = TRUE)
        n_cells_cl <- sum(clusters == cl)
        head(data.frame(
            cluster   = cl,
            type      = names(es_cl),
            scores    = es_cl / n_cells_cl,
            ncells    = n_cells_cl
        ), current_params$top_cell_types)
    }))

    # Best type per cluster
    best_types <- cL_results %>%
        dplyr::group_by(cluster) %>%
        dplyr::top_n(1, scores) %>%
        dplyr::mutate(
            type = ifelse(scores < 0, "Unknown", as.character(type))
        )

    # Map back to cells
    cell_types <- setNames(
        as.character(best_types$type[match(clusters, best_types$cluster)]),
        rownames(obj@meta.data)
    )

    score_col <- current_params$score_col
    obj@meta.data[[score_col]] <- cell_types

    # Per-cell cluster score
    cell_cluster_scores <- cL_results$scores[match(clusters, cL_results$cluster)]
    obj@meta.data[[paste0(score_col, ".score")]] <- cell_cluster_scores

    # Collect metrics
    n_assigned  <- length(unique(cell_types[cell_types != "Unknown"]))
    pct_unknown <- mean(cell_types == "Unknown") * 100
    mean_score  <- mean(cell_cluster_scores[cell_types != "Unknown"], na.rm = TRUE)
    top3        <- names(sort(table(cell_types[cell_types != "Unknown"]), decreasing = TRUE))[1:min(3L, n_assigned)]

    # Metadata Footprinting
    if (is.null(obj@misc$analysis_history)) obj@misc$analysis_history <- list()
    obj@misc$analysis_history <- c(obj@misc$analysis_history, list(list(
        skill_id  = "sctype_score",
        params    = list(
            tissue_type  = current_params$tissue_type,
            scaled_assay = current_params$scaled_assay,
            cluster_col  = cluster_col
        ),
        metrics   = list(
            n_assigned_types = as.integer(n_assigned),
            pct_unknown      = pct_unknown,
            mean_top_score   = mean_score,
            top3_types       = top3,
            n_clusters       = as.integer(length(unique(clusters))),
            n_cells          = as.integer(ncol(obj))
        ),
        timestamp = format(Sys.time(), "%Y-%m-%dT%H:%M:%S")
    )))

    if (is.character(input_data)) saveRDS(obj, input_data)
    return(obj)
}

File: scanpy_rank_genes_groups/execution.py
"""
scanpy_rank_genes_groups — Differential Expression / Marker Gene Detection

PURPOSE:
    Identifies genes that distinguish each cell cluster from all others (one-vs-rest).
    Essential for cell type annotation and biological interpretation of clusters.

INPUTS:
    input_data  : str | AnnData
    params_dict : dict (optional)

DEFAULTS:
    groupby='leiden', method='wilcoxon', n_genes=100, use_raw=None,
    pts=True, key_added='rank_genes_groups'

OUTPUTS:
    AnnData with adata.uns['rank_genes_groups'] and adata.uns['analysis_history']
"""
import scanpy as sc
import anndata as ad
import numpy as np
from datetime import datetime


def _to_serializable(obj):
    """Recursively convert numpy types to Python natives for h5ad compatibility."""
    if isinstance(obj, dict):
        return {str(k): _to_serializable(v) for k, v in obj.items()}
    if isinstance(obj, (list, tuple)):
        return [_to_serializable(i) for i in obj]
    if isinstance(obj, np.integer):   return int(obj)
    if isinstance(obj, np.floating):  return float(obj)
    if isinstance(obj, np.ndarray):   return obj.tolist()
    if isinstance(obj, np.bool_):     return bool(obj)
    return obj


def run_rank_genes_groups(input_data, params_dict=None, default_params=None):
    # Memory-First
    adata = input_data if isinstance(input_data, ad.AnnData) else sc.read(input_data)

    # Parameter Safety
    default_params = default_params or {
        'groupby': 'leiden',
        'method': 'wilcoxon',
        'n_genes': 100,
        'use_raw': None,
        'pts': True,
        'key_added': 'rank_genes_groups',
    }
    current_params = {**default_params, **(params_dict or {})}

    # Auto-detect groupby fallback
    groupby = current_params['groupby']
    if groupby not in adata.obs.columns:
        # Try common alternatives
        for alt in ['louvain', 'leiden', 'cluster', 'cell_type']:
            if alt in adata.obs.columns:
                groupby = alt
                break
        else:
            raise KeyError(
                f"Column '{current_params['groupby']}' not found in adata.obs. "
                f"Available: {list(adata.obs.columns)}. Run clustering first."
            )

    # Auto-detect use_raw
    use_raw = current_params['use_raw']
    if use_raw is None:
        use_raw = adata.raw is not None

    kwargs = {
        'groupby': groupby,
        'method': current_params['method'],
        'n_genes': current_params['n_genes'],
        'use_raw': use_raw,
        'pts': current_params['pts'],
        'key_added': current_params['key_added'],
    }

    sc.tl.rank_genes_groups(adata, **kwargs)

    # Collect metrics
    key = current_params['key_added']
    result = adata.uns[key]
    groups = list(result['names'].dtype.names)
    n_groups = len(groups)

    # Top-1 score per group
    top1_scores = []
    for g in groups:
        scores = result['scores'][g]
        if len(scores) > 0:
            top1_scores.append(float(scores[0]))

    mean_top1_score = float(np.mean(top1_scores)) if top1_scores else 0.0
    min_top1_score = float(np.min(top1_scores)) if top1_scores else 0.0
    mean_genes_per_group = float(np.mean([
        len(result['names'][g]) for g in groups
    ]))

    # Metadata Footprinting
    if 'analysis_history' not in adata.uns:
        adata.uns['analysis_history'] = []
    adata.uns['analysis_history'].append(_to_serializable({
        'skill_id': 'scanpy_rank_genes_groups',
        'params': {**current_params, 'groupby': groupby, 'use_raw': use_raw},
        'metrics': {
            'n_groups': n_groups,
            'mean_top1_score': mean_top1_score,
            'min_top1_score': min_top1_score,
            'mean_genes_per_group': mean_genes_per_group,
        },
        'timestamp': datetime.now().isoformat()
    }))

    if not isinstance(input_data, ad.AnnData):
        adata.write(input_data)
    return adata
